docs/Backend_tutorials/features/graphql.md

```diff
@@ -1,5 +1,139 @@
+
+### Inputs (arguments)
+https://graphql-ruby.org/type_definitions/input_objects.html
+
+Inputs are subset of types. They are used for defining arguments of the type. We are using `argument` method for defining arguments of the type. In Easy we are using `Types::BaseInput` for all our inputs. This class is defined in `app/api/easy_graphql/types/base_input.rb`. Input represents a hash of *entity* attributes, primarily for `Mutations`.
+
+Please inputs into `app/api/easy_graphql/types/inputs` directory. Name convention of input is `model_name_attributes.rb`.
+
+> [!WARNING]
+> `_attrubtes` suffix is used to distinguish input from type.
+
+## Resolvers
+https://graphql-ruby.org/fields/resolvers.html
+
+Resolvers are used for defining how to fetch the data. We are using `argument` method for defining params of the type. We are using `resolve` method for defining how to fetch the data.
+
+We are using `Resolvers::Base` for all our resolvers. This class is defined in `app/api/easy_graphql/resolvers/base.rb`.
+
+Place resolvers into `app/api/easy_graphql/resolvers` directory. Name convention of resolver should describe "what resolve does" OR if its fetch entity, just use entity name in plural. For example `Issues` or `MarkAsRead`.
+
+### Best-practice
+
+1. **Use `resolve` method for fetching data.**
+2. **Use `type` method for defining return type.**
+3. **Use `argument` method for defining fields of the type.** Do not forget specify `required` option for argument.
+4. If you find specify entity, use `find` method for fetching entity. It raise `ActiveRecord::RecordNotFound` if entity not found = This is correct.
+5. Do not forget to check permissions for visibility.
+6. Prefer use `arguments` method for accessing arguments.
+
+#### Example
+
+```ruby title="example usage of resolver" lineNumbers
+module EasyGraphql
+  module Resolvers
+    class IssueDetail < Resolvers::Base
+      description "Return Issue object based on given ID."
+
+      type Types::Issue, null: false
+
+      argument :id, ID, required: true
+
+      def resolve(**_args)
+        { issue: }
+      end
+
+      private
+      
+      def issue
+        @issue ||= ::Issue.preload(journals: :details).visible.find(arguments[:id])
+      end
+
+    end
+  end
+end
+```
+
+## Mutations
+https://graphql-ruby.org/mutations/mutation_classes.html
+
+Mutations are used for defining how to change the data. We are using `argument` method for defining arguments of the type. We are using `resolve` method for defining how to change the data. We are using `field` method for defining fields of the type (return data of mutation).
+
+We are using `Mutations::Base` for all our mutations. This class is defined in `app/api/easy_graphql/mutations/base.rb`.
+
+Place mutations into `app/api/easy_graphql/mutations` directory. Name convention of mutation should describe "what mutation does" OR if its change entity, just use entity name **with action suffix**. For example `IssueCreate` or `IssueUpdate`.
```

Comment:
```suggestion
Place mutations into `app/api/easy_graphql/mutations/plural_model_name/action` directory. Name convention of mutation should describe "what mutation does". For example `app/api/easy_graphql/mutations/issues/create`
```
Same as services.

Reply:
EXAMPLE 

```ruby
module EasyGraphql
  module Mutations
    module EasyIssueTemplates
      class Create < Base
        description "Create EasyIssueTemplate"

        argument :attributes, Types::Inputs::EasyIssueTemplateAttributes, required: true

        field :easy_issue_template, Types::EasyIssueTemplate, null: true
        field :errors, [Types::Error], null: true

        def resolve(**_args)
          if easy_issue_template.save
            { easy_issue_template:, errors: [] }
          else
            errors.add_entity_errors(easy_issue_template)
            { errors: }
          end
        end

        private

        def easy_issue_template
          @easy_issue_template ||= ::EasyIssueTemplate.new(arguments[:attributes].to_h)
        end

      end
    end
  end
end
```

Reply:
Nope, we don't want use namespaces. It's seems not good practice. It's too complicated to read. Plus use `prefix` is consistent with FE usage.